src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+	}
+	start.min = entry.cdte_addr.msf.minute;
+	start.sec = entry.cdte_addr.msf.second;
+	start.fr  = entry.cdte_addr.msf.frame;
+	attr      = (entry.cdte_ctrl << 4) | entry.cdte_adr;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: GetAudioTrackInfo for track %u => "
+	         "MSF %02d:%02d:%02d, which is sector %d",
+	         track,
+	         start.min,
+	         start.sec,
+	         start.fr,
+	         msf_to_frames(start));
+#endif
+	return true;
+}
+
+bool CDROM_Interface_Ioctl::GetAudioSub(unsigned char& attr, unsigned char& track,
+                                        unsigned char& index, TMSF& relPos,
+                                        TMSF& absPos)
+{
+	if (cdrom_fd == -1) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioSub: cdrom_fd not open");
+#endif
+		return false;
+	}
+	cdrom_subchnl sub = {};
+	sub.cdsc_format   = CDROM_MSF;
+	if (ioctl(cdrom_fd, CDROMSUBCHNL, &sub) != 0) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioSub: ioctl CDROMSUBCHNL failed");
+#endif
+		return false;
+	}
+	attr       = (sub.cdsc_ctrl << 4) | sub.cdsc_adr;
+	track      = sub.cdsc_trk;
+	index      = sub.cdsc_ind;
+	relPos.min = sub.cdsc_reladdr.msf.minute;
+	relPos.sec = sub.cdsc_reladdr.msf.second;
+	relPos.fr  = sub.cdsc_reladdr.msf.frame;
+	absPos.min = sub.cdsc_absaddr.msf.minute;
+	absPos.sec = sub.cdsc_absaddr.msf.second;
+	absPos.fr  = sub.cdsc_absaddr.msf.frame;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: GetAudioSub => position at %02d:%02d:%02d (on sector %u) "
+	         "within track %u at %02d:%02d:%02d (at its sector %u)",
+	         absPos.min,
+	         absPos.sec,
+	         absPos.fr,
+	         msf_to_frames(absPos),
+	         track,
+	         relPos.min,
+	         relPos.sec,
+	         relPos.fr,
+	         msf_to_frames(relPos));
+#endif
+	return true;
+}
+
+bool CDROM_Interface_Ioctl::GetAudioStatus(bool& playing, bool& pause)
+{
+	playing = this->playing;
+	pause   = this->paused;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: GetAudioStatus => %s and %s",
+	         playing ? "is playing" : "stopped",
+	         pause ? "paused" : "not paused");
+#endif
+	return true;
+}
+
+// Called from CMscdex::GetDeviceStatus - dos_mscdex.cpp line 821
+// Function doesn't check return value or initialize variables so just set some
+// defaults and never fail.
+bool CDROM_Interface_Ioctl::GetMediaTrayStatus(bool& mediaPresent,
+                                               bool& mediaChanged, bool& trayOpen)
+{
+	mediaPresent = false;
+	mediaChanged = false;
+	trayOpen     = false;
+	if (cdrom_fd == -1) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetMediaTrayStatus: cdrom_fd not open");
+#endif
+		return true;
+	}
+	switch (ioctl(cdrom_fd, CDROM_DRIVE_STATUS, CDSL_CURRENT)) {
+	case CDS_TRAY_OPEN: trayOpen = true; break;
+	case CDS_DISC_OK: mediaPresent = true; break;
+	}
+	int ret = ioctl(cdrom_fd, CDROM_MEDIA_CHANGED, CDSL_CURRENT);
+	mediaChanged = (ret > 0) && (ret & 1);
+#ifdef DEBUG_IOCTL
```

Comment:
Blank above.
(Stopping now... you get the picture 😄 )

Reply:
Fixed